packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts

```diff
@@ -5,37 +5,129 @@
+  message: string;
+  name: string;
+  result: Result[];
+  stack: string;
+  status: number;
+  warnings: any[];
+}
+
+// move this elsewhere?
+vscode.workspace.onDidChangeTextDocument((e => {
+  if (ForceSourceDeployExecutor._errorCollection.has(e.document.uri)) {
+    ForceSourceDeployExecutor._errorCollection.delete(e.document.uri);
+  }
+}));
+
+export class ForceSourceDeployExecutor extends SfdxCommandletExecutor<SelectedPath> {
+
+  public static _errorCollection = vscode.languages.createDiagnosticCollection('deploy-errors');
+
   public build(data: SelectedPath): Command {
     const commandBuilder = new SfdxCommandBuilder()
       .withDescription(nls.localize('force_source_deploy_text'))
       .withArg('force:source:deploy')
-      .withLogName('force_source_deploy');
+      .withLogName('force_source_deploy')
+      .withJson();
     if (data.type === FileType.Manifest) {
       commandBuilder.withFlag('--manifest', data.filePath);
     } else {
       commandBuilder.withFlag('--sourcepath', data.filePath);
     }
     return commandBuilder.build();
   }
+
+  public execute(response: ContinueResponse<SelectedPath>): void {
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: vscode.workspace.rootPath
+    }).execute(cancellationToken);
+
+    let stdErr = '';
+    execution.stderrSubject.subscribe(realData => {
+      stdErr += realData.toString();
+    });
+
+    execution.processExitSubject.subscribe(async exitCode => {
+      if (exitCode !== 0) {
+        try {
+          const compileErrs = this.getDeployResultData(stdErr);
+          const fileErrors = this.groupErrorsByPath(compileErrs);
+          Object.keys(fileErrors).forEach(filePath => {
+            const uri = vscode.Uri.file(path.join(vscode.workspace.rootPath || '', filePath));
+            const fileDiagnostics: vscode.Diagnostic[] = [];
+            fileErrors[filePath].forEach(err => {
+              fileDiagnostics.push(this.createDiagnosticFromResult(err));
+            });
+            ForceSourceDeployExecutor._errorCollection.set(uri, fileDiagnostics);
+          });
+        } catch (e) {
+          // something else happened...?
+          console.log(`Could not parse Compile Errors`);
+        }
+      }
+    });
+
+    notificationService.reportExecutionError(
+      execution.command.toString(),
+      (execution.stderrSubject as any) as Observable<Error | undefined>
+    );
+    this.logMetric(execution.command.logName);
+    channelService.showChannelOutput();
+    channelService.streamCommandOutput(execution);
+    ProgressNotification.show(execution, cancellationTokenSource);
+    taskViewService.addCommandExecution(execution, cancellationTokenSource);
+  }
+
+  private getDeployResultData(stdErr: string) {
+    const stdErrLines = stdErr.split('\n');
```

Comment:
Is this safe on Windows?

Reply:
had the same thought... I'm not sure.  Regardless I should use `require('os').EOL` instead

Reply:
@lcampos @vazexqi  What are your thoughts about the line below (for some reason github won't let me start a new comment)?

There's gotta be a better way of handling the update + other junk before the response json.  IMO, it would be easier if the error results were in stdout... 

How have you guys handled this in other commands?